File: POO/Constructores_Clases/Abstract class/JAVA/Animal.java
public abstract class Animal {
    protected String nombre;
    protected String especie;

    // Constructor sin parámetros
    public Animal() {
        this.nombre = "Sin nombre";
        this.especie = "Desconocida";
    }

    // Métodos abstractos
    public abstract String hacerSonido();
    public abstract String moverse();

    public void mostrar() {
        System.out.println(nombre + " - " + especie);
    }
}

class Gato extends Animal {
    public Gato() {
        super();
        this.nombre = "Michi";
        this.especie = "Gato";
    }

    @Override
    public String hacerSonido() {
        return "Miau miau!";
    }

    @Override
    public String moverse() {
        return "Caminando sigilosamente";
    }

    @Override
    public void mostrar() {
        super.mostrar();
        System.out.println("Sonido: " + hacerSonido());
        System.out.println("Movimiento: " + moverse());
    }

    public static void main(String[] args) {
        Animal gato = new Gato();
        gato.mostrar();
    }
}

File: POO/Constructores_Clases/Abstract class/js/Animal.js
// Simulación de clase abstracta en JavaScript
class Animal {
    constructor() {
        if (this.constructor === Animal) {
            throw new Error("No se puede instanciar una clase abstracta");
        }
        this.nombre = "Sin nombre";
        this.especie = "Desconocida";
    }

    // Métodos abstractos
    hacerSonido() {
        throw new Error("Método abstracto debe ser implementado");
    }

    moverse() {
        throw new Error("Método abstracto debe ser implementado");
    }

    mostrar() {
        console.log(`${this.nombre} - ${this.especie}`);
    }
}

class Perro extends Animal {
    constructor() {
        super();
        this.nombre = "Rex";
        this.especie = "Perro";
    }

    hacerSonido() {
        return "Guau guau!";
    }

    moverse() {
        return "Corriendo en cuatro patas";
    }

    mostrar() {
        super.mostrar();
        console.log(`Sonido: ${this.hacerSonido()}`);
        console.log(`Movimiento: ${this.moverse()}`);
    }
}

// Uso
const perro = new Perro();
perro.mostrar();